### PR TITLE
rshim for Nvidia BlueField

### PR DESCRIPTION
**Pull request by nikstur**
###### Description of changes

Add the [rshim-user-space](https://github.com/Mellanox/rshim-user-space/tree/master) package and a corresponding module.
These are needed to interact with a Nvidia/Mellanox BlueField DPU/SmartNic.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/rshim-user-space/default.nix [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+  preConfigure = "./bootstrap.sh";
+
+  installPhase = ''
+    mkdir -p "$out"/bin
+    cp -a src/rshim "$out"/bin/
+  '';
+
+  meta = with lib; {
+    description = "user-space rshim driver for the BlueField SoC";
```
Comment (RaitoBezarius) [inline pkgs/tools/misc/rshim-user-space/default.nix:47]:
```suggestion
    description = "User-space rshim driver for the BlueField SoC";
```

File: nixos/modules/services/misc/rshim.nix [hunk 2]
```diff
@@ -0,0 +1,97 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.rshim;
+
+  rshimCommand = [ "${cfg.package}/bin/rshim" ]
```
Comment (RaitoBezarius) [inline nixos/modules/services/misc/rshim.nix:6]: you should use `escapeShellArguments`

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]

File: nixos/modules/services/misc/rshim.nix (hunk 2 above)
Comment (nikstur) [reply nixos/modules/services/misc/rshim.nix]: I shall

File: nixos/modules/services/misc/rshim.nix [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      '';
+      default = null;
+      example = "pcie-04:00.2";
+    };
+
+    index = lib.mkOption {
+      type = with lib.types; nullOr int;
+      description = lib.mdDoc ''
+        Specify the index to create device path /dev/rshim<index>. It's also
+        used to create network interface name tmfifo_net<index>. This option is
```
Comment (lf-) [inline nixos/modules/services/misc/rshim.nix:44]: do you need backslash escaping here? I think this name should be in backticks to make sure there's no weird html parsing.

File: nixos/modules/services/misc/rshim.nix [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        cfg.config;
+    };
+
+    systemd.services.rshim = {
+      after = [ "network.target" ];
+      serviceConfig = {
+        Restart = "always";
+        Type = "forking";
+        ExecStart = [
+          (lib.concatStringsSep " \\\n" rshimCommand)
```
Comment (lf-) [inline nixos/modules/services/misc/rshim.nix:90]: It seems a little odd to put \n here (how does systemd interpret this? it's not using typical shell semantics).

**Review by lf- — COMMENTED** [review]
This looks like an excellently written module, great work! I noted some extremely tiny quibbles.

File: nixos/modules/services/misc/rshim.nix (hunk 3 above)
Comment (nikstur) [reply nixos/modules/services/misc/rshim.nix:44]: Yes you're right. Added it.

File: nixos/modules/services/misc/rshim.nix (hunk 4 above)
Comment (nikstur) [reply nixos/modules/services/misc/rshim.nix:90]: > Lines ending in a backslash are concatenated with the following line while reading and the backslash is replaced by a space character

from [systemd.syntax](https://www.freedesktop.org/software/systemd/man/systemd.syntax.html#).

I think for debugging by looking at the generated unit file, it's nice to have human-readable line breaks. However from a parsing perspective it makes no difference.

**Comment by GovanifY** [comment]
Don't really have anything more to add than the resolved comments from a cursory look, although I didn't test it. 
Will try to do so whenever I get some free time, thanks for the work!

**Review by RaitoBezarius — APPROVED** [review]

**Comment by nikstur** [comment]
> Don't really have anything more to add than the resolved comments from a cursory look, although I didn't test it. Will try to do so whenever I get some free time, thanks for the work!

Do you have access to BlueFields? I just tested the module with a system with two BlueFields and it does what it's supposed to. Please report back if there's anything to fix or improve.

**Comment by GovanifY** [comment]
I can borrow one yeah; if it all works then it’s fine, thanks!

**Pull request merged**